exonum/src/blockchain/builder.rs

```diff
@@ -0,0 +1,213 @@
+        Self {
+            database: database.into(),
+            genesis_config,
+            service_keypair,
+            runtimes: Vec::new(),
+            builtin_instances: Vec::new(),
+        }
+    }
+
+    pub fn with_rust_runtime(
+        mut self,
+        services: impl IntoIterator<Item = ServiceInstances>,
+    ) -> Self {
+        let mut runtime = RustRuntime::new();
+        for service in services {
+            runtime.add_service_factory(service.factory);
+            self.builtin_instances.extend(service.instances);
+        }
+        self.runtimes.push(runtime.into());
+        self
+    }
+
+    /// Returns blockchain instance, creates and commits the genesis block with the specified
+    /// genesis configuration if the blockchain has not been initialized.
+    /// Otherwise restores dispatcher state from database.
+    ///
+    /// # Panics
+    ///
+    /// * If the genesis block was not committed.
+    /// * If storage version is not specified or not supported.
+    pub fn finalize(
+        self,
+        api_sender: ApiSender,
+        dispatcher_requests: mpsc::Sender<InternalRequest>,
+    ) -> Result<Blockchain, failure::Error> {
+        let mut dispatcher = Dispatcher::with_runtimes(self.runtimes, dispatcher_requests);
+        // If genesis block had been already created just restores dispatcher state from database
+        // otherwise creates genesis block with the given specification.
+        let has_genesis_block = {
+            let snapshot = self.database.snapshot();
+            !Schema::new(snapshot.as_ref())
+                .block_hashes_by_height()
+                .is_empty()
+        };
+
+        Ok(if has_genesis_block {
+            let snapshot = self.database.snapshot();
+            dispatcher.restore_state(snapshot.as_ref());
+            Blockchain::with_dispatcher(
+                self.database,
+                dispatcher,
+                self.service_keypair.0,
+                self.service_keypair.1,
+                api_sender,
+            )
+        } else {
+            // Creates blockchain with the new genesis block.
+            let mut blockchain = Blockchain::with_dispatcher(
+                self.database,
+                dispatcher,
+                self.service_keypair.0,
+                self.service_keypair.1,
+                api_sender,
+            );
+            // Adds builtin services.
+            blockchain.merge({
+                let fork = blockchain.fork();
+                self.builtin_instances.into_iter().for_each(|service| {
+                    blockchain
+                        .dispatcher()
+                        .add_builtin_service(&fork, service.0, service.1)
+                });
+                fork.into_patch()
+            })?;
+            // Commits genesis block.
+            blockchain.create_genesis_block(self.genesis_config)?;
+            blockchain
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct ServiceInstances {
```

Comment:
`ServiceInstances` is a little bit weird as for me. What about `ServiceInstancesContainer` or `ServiceInstancesHolder` or something like this but in the singular ?